rust/datafusion/src/execution/physical_plan/limit.rs

```diff
@@ -0,0 +1,208 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines the LIMIT plan
+
+use crate::error::{ExecutionError, Result};
+use crate::execution::physical_plan::common::RecordBatchIterator;
+use crate::execution::physical_plan::ExecutionPlan;
+use crate::execution::physical_plan::{BatchIterator, Partition};
+use arrow::array::ArrayRef;
+use arrow::compute::array_ops::limit;
+use arrow::datatypes::Schema;
+use arrow::record_batch::RecordBatch;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::thread::JoinHandle;
+
+/// Limit execution plan
+pub struct LimitExec {
+    /// Input schema
+    schema: Arc<Schema>,
+    /// Input partitions
+    partitions: Vec<Arc<dyn Partition>>,
+    /// Maximum number of rows to return
+    limit: usize,
+}
+
+impl LimitExec {
+    /// Create a new MergeExec
+    pub fn new(
+        schema: Arc<Schema>,
+        partitions: Vec<Arc<dyn Partition>>,
+        limit: usize,
+    ) -> Self {
+        LimitExec {
+            schema,
+            partitions,
+            limit,
+        }
+    }
+}
+
+impl ExecutionPlan for LimitExec {
+    fn schema(&self) -> Arc<Schema> {
+        self.schema.clone()
+    }
+
+    fn partitions(&self) -> Result<Vec<Arc<dyn Partition>>> {
+        Ok(vec![Arc::new(LimitPartition {
+            schema: self.schema.clone(),
+            partitions: self.partitions.clone(),
+            limit: self.limit,
+        })])
+    }
+}
+
+struct LimitPartition {
+    /// Input schema
+    schema: Arc<Schema>,
+    /// Input partitions
+    partitions: Vec<Arc<dyn Partition>>,
+    /// Maximum number of rows to return
+    limit: usize,
+}
+
+impl Partition for LimitPartition {
+    fn execute(&self) -> Result<Arc<Mutex<dyn BatchIterator>>> {
+        // collect up to "limit" rows on each partition
+        let threads: Vec<JoinHandle<Result<Vec<RecordBatch>>>> = self
+            .partitions
+            .iter()
```

Comment:
Would be an overkill to use rayon's par_iter()?

Reply:
For this round my plan was to keep things simple and do a second round of optimizations that I've started tracking in https://jira.apache.org/jira/browse/ARROW-6689 and in particular see the subtask https://jira.apache.org/jira/browse/ARROW-6691 related to threading.

Reply:
I think that's a "no" for rayon :D